enhancements/authentication/service-ca-cert-generation-for-statefulset-pods.md

```diff
@@ -0,0 +1,327 @@
+
+If a client wants to connect to _only_ a “true localhost” service,
+a service CA-signed certificate does not provide that property.
+The best way to do that is to do it directly inside the process,
+without using TLS (and the associated cryptography overhead) at all.
+Second best is to rely on known properties of the environment,
+i.e. to assume that 127.0.0.1 is localhost-only
+and to use a non-TLS connection on that address.
+
+If a _pod_ had to use TLS for localhost for some reason, the closest to a right way to do that
+(relying on TLS to verify that only loopback connections happen) is to:
+- Create a temporary CA (specific to the individual pod for)
+- Generate, and have that CA sign a localhost certificate (specific to the individual pod)
+- Set up the CA certificate as a trusted root
+- _Irreversibly erase_ the CA private key
+- Set up the pod with the generated localhost certificate+private key.
+- (This doesn’t benefit from service-ca involvement at all.)
+
+Similarly single-purpose CA could be set up for NodeIP services
+(if it makes sense to connect to node IP addresses directly at all)
+so that the client connects to the desired NodeIP service
+instead of to any NodeIP service.
+
+### Risks and Mitigations
+
+The only mitigations for the risks below are not to use this feature,
+(and prevent untrusted users from using it — using an admission controller?)
+so that the wildcard certificate is not created at all,
+or perhaps $somehow ensure that any users/code with access to the wildcard certificate
+is not likely to be compromised.
+
+#### Unexpected DNS Collisions in the Future
+
+https://github.com/kubernetes/dns/blob/master/docs/specification.md
+only specifies what DNS records must exist, not what must not exist,
+so the generated wildcard certificate could match unwanted DNS entries.
+
+The DNS specification does not even guarantee that different namespaces
+will use different DNS subtrees.
+
+Given the existing naming conventions of in-cluster DNS,
+such unexpected collisions seem somewhat unlikely,
+but can’t be ruled out in advance.
+
+#### No Individual Pod Identity
+
+The wildcard certificate does not protect the intra-StatefulSet communication
+against re-routing to unintended members of that StatefulSet
+(or to a different StatefulSet attached to the same headless service).
+
+In addition, any application that uses a different set of certificates to achieve that goal
+**must not accept the Service CA as a valid root of trust for connections intended
+for individual StatefulSet pods**.
+
+## Design Details
+
+### Open Questions
+
+N/A
+
+### Test Plan
+
+Unit and e2e tests similar to the existing service certificate tests, both in structure and scope.
+
+Ensure that upgrades (or alternatively any controller re-syncs)
+cause re-generation of certificates for headless services
+to include the wildcard subjects, if they are missing.
+
+### Graduation Criteria
```

Comment:
Given the reduced scope of implementation, these sections are probably no longer relevant. The existing annotation will just gain support for generating wildcard certificates for headless services.

Reply:
The one thing that can’t just be a part of a service-CA PR is the end-user documentation. I realize that just listing it here won’t make it happen :)

I’ll drop the rest.